src/SDKs/Azure.Base/data-plane/Azure.Base/Http/Pipeline/RetryPolicy.cs

```diff
@@ -14,14 +16,48 @@ public static RetryPolicy CreateFixed(int maxRetries, TimeSpan delay, params int
         public override async Task ProcessAsync(HttpPipelineMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
         {
             int attempt = 1;
+            List<Exception> exceptions = null;
             while (true)
             {
-                await ProcessNextAsync(pipeline, message).ConfigureAwait(false);
-                if (!ShouldRetry(message, attempt++, out var delay)) return;
-                if (delay > TimeSpan.Zero) await Task.Delay(delay, message.Cancellation).ConfigureAwait(false);
+                Exception lastException = null;
+
+                try
+                {
+                    await ProcessNextAsync(pipeline, message).ConfigureAwait(false);
+                }
+                catch (Exception ex)
+                {
+                    if (exceptions == null)
+                    {
+                        exceptions = new List<Exception>();
+                    }
+                    exceptions.Add(ex);
+
+                    lastException = ex;
+                }
+
+                if (!ShouldRetry(message, lastException, attempt++, out var delay))
+                {
+                    if (lastException != null)
+                    {
+                        throw new AggregateException($"Retry failed after {attempt - 1} tries.", exceptions);
+                    }
+
+                    return;
+                }
+
+                if (delay > TimeSpan.Zero)
+                {
+                    await Delay(delay, message.Cancellation);
+                }
             }
         }
 
-        protected abstract bool ShouldRetry(HttpPipelineMessage message, int attempted, out TimeSpan delay);
+        protected virtual async Task Delay(TimeSpan time, CancellationToken cancellationToken)
```

Comment:
Why is this virtual?

Reply:
So I can override it in a test and avoid writing flaky and slow tests.

Reply:
Why not set the delay to 0 when run in tests?

Reply:
How do I know that it tried to wait using the correct delay value?

Reply:
Especially when writing tests for exponential retry knowing the exact value is important.

Reply:
Call ShouldRetry in your tests.

Reply:
This way I'm testing end to end using the pipeline. Why are you worried about virtual? I can make it internal too.

Reply:
If you can make it internal, then great! I worry about unnecessary extensibility points constraining how we can evolve the APIs in the future.